### PR TITLE
Implement generic Chisel module, remove multiplier, and Xilinx Vivado synthesis

### PR DESCRIPTION
**Pull request by edwardcwang**
@colinschmidt Pull request as we discussed today to keep bar up to sync on orthogonal changes.

### CODE REVIEW

File: src/addons/synthesis/vivado/generate-macros [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    >&2 echo "You must set \$dcp to dump .dcp files";
+    exit 1;
+fi
+
+# Clean macro file just in case
+rm -f "$output"
+
+# Gimme my SRAMs, yum yum.
+#$script_dir/generate-macro --sram-module "_plsi_tech_vivado_prebuilt_sram_32b_4096" --width 32 --depth 4096 -o "${workdir}/32b_4096.v" "$vivado" --board_files "$board_files" --dcp "${dcp}/32b_4096.dcp"
+$script_dir/generate-macro --sram-module "_plsi_tech_vivado_prebuilt_sram_32b_2048" --width 32 --depth 2048 -o "${workdir}/32b_2048.v" "$vivado"  --board_files "$board_files" --dcp "${dcp}/32b_2048.dcp"
```
Comment (colinschmidt) [inline src/addons/synthesis/vivado/generate-macros]: Shouldn't the srams you want be specified outside of this file?

File: src/technologies/vivado.tech.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  "name": "FPGA block memories generated by Xilinx Vivado",
+  "installs_commented_out": [
+    {
+      "path": "tsmc28",
+      "base var": "TSMC_INSTALL_DIR"
+    }
+  ],
+  "libraries": [
+    {
+      "liberty file": "tsmc28/IP/tcbn28hpmbwp35_111a/TSMCHOME/digital/Back_End/Front_End/timing_power_noise/NLDM/tcbn28hpmbwp35_111a/tcbn28hpmbwp35tt0p9v25c.lib",
```
Comment (colinschmidt) [inline src/technologies/vivado.tech.json]: This doesn't seem right.

**Comment by edwardcwang** [comment]
As we discussed briefly today, I will tag the comments as TODOs in the code and merge.

File: src/addons/core-generator/chisel/pinmap.format.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Format for the pinmap.json file.
```
Comment (edwardcwang) [inline src/addons/core-generator/chisel/pinmap.format.json]: That is right. I've added a note to the file.

File: src/addons/soc-generator/vivado/generate-system-v [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                    raise ValueError("Trying to assign pin " + pinname + " twice")
+                pinmap[pinname] = PinmapEntry(dut_name=entry['dut_name'], pinname=pinname, type=typ, dut_name_index=-1)
+    return pinmap, pinmap_arrays
+
+def main_func(args):
+    top_module = args.top
+
+    pins = read_pin_database(args.database)
+    pinmap, pinmap_arrays = read_pin_map(args.pinmap)
+    print(pinmap)
```
Comment (edwardcwang) [inline src/addons/soc-generator/vivado/generate-system-v]: Good point. I will just comment them out and add a TODO for now.

File: src/addons/soc-generator/vivado/generate-system-v [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("-p", "--pinmap", required=True, help="Path to the pinout in the format of pinout.format.json.")
+    parser.add_argument("-d", "--database", required=True, help="Path to the pin database. Format is \"Comment?,Pin name,Type\" (see the csv file itself).")
+    parser.add_argument("-t", "--template", required=True, help="Path to system.template.v.")
+    parser.add_argument("--top", required=True, help="Module name of the top-level module.")
+    parser.add_argument("-o", "--output", required=True, help="Output path to store the generated system.v file.")
+
+    sys.exit(main_func(parser.parse_args()))
```
Comment (edwardcwang) [inline src/addons/soc-generator/vivado/generate-system-v]: Agreed, let's talk about this at our next PLSI meeting.

File: src/addons/soc-generator/vivado/pins.csv [hunk 6]
```diff
@@ -0,0 +1,25 @@
+Comment?,Pin name,Type
+Comment,Name of the pin,“input” or “output”
```
Comment (edwardcwang) [inline src/addons/soc-generator/vivado/pins.csv]: Agreed, will do so.

File: src/addons/synthesis/vivado/generate-macros (hunk 1 above)
Comment (edwardcwang) [inline src/addons/synthesis/vivado/generate-macros]: For now in order to keep things simple, we simply generate a (small at the moment) list of SRAMs and let the PLSI memory compiler figure it out from there. In the future we could look into custom generating (couldn't figure out the depends script thing).

File: src/technologies/vivado.tech.json (hunk 2 above)
Comment (edwardcwang) [inline src/technologies/vivado.tech.json]: I copied these from the tsmc28 but it seems like they can just be removed if not using dc/icc. I'll go ahead and remove them.

**Comment by edwardcwang** [comment]
Thanks for the review; merged. I added the discussion items to be tabled for the next meeting.

**Pull request merged**